api/core/model_runtime/model_providers/vertex_ai/llm/llm.py

```diff
@@ -0,0 +1,446 @@
+        :param credentials: credentials
+        :param response: response
+        :param prompt_messages: prompt messages
+        :return: llm response chunk generator result
+        """
+        print("response");
+        print(response)
+        index = -1
+        for chunk in response:
+            print("chunk")
+            print(chunk)
+            for part in chunk.candidates[0].content.parts:
+                print("part")
+                print(part)
+                assistant_prompt_message = AssistantPromptMessage(
+                    content=''
+                )
+
+                if part.text:
+                    assistant_prompt_message.content += part.text
+
+                if part.function_call:
+                    assistant_prompt_message.tool_calls = [
+                        AssistantPromptMessage.ToolCall(
+                            id=part.function_call.name,
+                            type='function',
+                            function=AssistantPromptMessage.ToolCall.ToolCallFunction(
+                                name=part.function_call.name,
+                                arguments=json.dumps({
+                                    key: value 
+                                    for key, value in part.function_call.args.items()
+                                })
+                            )
+                        )
+                    ]
+
+                index += 1
+    
+                if not hasattr(chunk, 'finish_reason') or not chunk.finish_reason:                    
+                    # transform assistant message to prompt message
+                    yield LLMResultChunk(
+                        model=model,
+                        prompt_messages=prompt_messages,
+                        delta=LLMResultChunkDelta(
+                            index=index,
+                            message=assistant_prompt_message
+                        )
+                    )
+                else:
+                    
+                    # calculate num tokens
+                    prompt_tokens = self.get_num_tokens(model, credentials, prompt_messages)
+                    completion_tokens = self.get_num_tokens(model, credentials, [assistant_prompt_message])
+
+                    # transform usage
+                    usage = self._calc_response_usage(model, credentials, prompt_tokens, completion_tokens)
+                    
+                    yield LLMResultChunk(
+                        model=model,
+                        prompt_messages=prompt_messages,
+                        delta=LLMResultChunkDelta(
+                            index=index,
+                            message=assistant_prompt_message,
+                            finish_reason=chunk.candidates[0].finish_reason,
+                            usage=usage
+                        )
+                    )
+
+    def _convert_one_message_to_text(self, message: PromptMessage) -> str:
```

Comment:
What about let this class inherit from `GoogleLargeLanguageModel` so that lots of codes can be reused, such as `_convert_one_message_to_text`, `handle_generate_stream_response` etc.

Reply:
Well I did not want to tie them together as something change in `GoogleLargeLanguageModel` not must change in here. Vertex AI library and Google AI are two separate libraries developed in two different road maps. I know that there might me common function not depending on library but still. 

Basically there are many common functions across almost all models but they are not much connected. 

What do you think?

Reply:
Sure, I agree with you, that makes sense. I just wanted to point out that they use many of the same methods, But well, it's just a small suggestion :)